### PR TITLE
REST API: Check role eligibility after authentication without WPCom

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #8453

## Description

This PR adds a check for role eligibility after authentication with site credentials. Changes include:
- Enabled application password for user information check. This makes it easier to reuse the existing `UserRemote`.
  - Updated `UserMapper` to parse data without the `data` envelope.
  - Enabled the option to convert the user information fetch to a REST request with `availableAsRESTRequest`.
  - Removed empty components when constructing REST requests to avoid redundant `/` causing invalid URLs.
- Added role eligibility check in `AuthenticationManager`:
  - If the check fails, display a generic alert.
  - If the check returns a user without sufficient permission, display the ineligible role screen.
  - If the user is either an administrator or shop manager, show the home screen.  

## Testing instructions

**1. With insufficient role:**
Prerequisite: Create a non-admin/shop manager user in your test site and make sure that application password is enabled.
- Enable the feature flag `applicationPasswordAuthenticationForSiteCredentialLogin` and build the app.
- Log out of the app if needed.
- On the prologue screen, select "Enter your site address". If this option is not available, [disable simplified login](https://github.com/woocommerce/woocommerce-ios/blob/8d471f969911067311412d72f58de6ae2c966c6a/WooCommerce/Classes/Authentication/AuthenticationManager.swift#L61) and build the app again.
- Enter your test store address and select "Sign in with site credentials".
- Enter correct credentials and select "Continue". Notice that after a while, a screen for insufficient role should be displayed.

**2. With issue fetching user informatino:**
Prerequisite: Make sure that application password is enabled on your site.
- Enable the feature flag `applicationPasswordAuthenticationForSiteCredentialLogin` and build the app.
- Log out of the app if needed.
- On the prologue screen, select "Enter your site address". If this option is not available, [disable simplified login](https://github.com/woocommerce/woocommerce-ios/blob/8d471f969911067311412d72f58de6ae2c966c6a/WooCommerce/Classes/Authentication/AuthenticationManager.swift#L61) and build the app again.
- Enter your test store address and select "Sign in with site credentials".
- Enter correct credentials and select "Continue".
- Use Proxyman to intercept the user fetch request (/wp/v2/users/me) or disable your network connection when this request is made.
- Notice that an alert is displayed saying application password cannot be generated.
- Tap the "Retry" button on the alert, notice that the alert is dismissed. If the issue persists, the same alert should be presented again.
- Tap the "Log In With Another Account" button. Notice that you are navigated to the prologue screen to log in again.

## Screenshots

| Ineligibility | Role check failure |
| ----- | ----- |
|  |  |

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/NetworkingTests/Responses/user-complete-without-data.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "data": {
```
Comment (selanthiraiyan) [inline Networking/NetworkingTests/Responses/user-complete-without-data.json]: I think we have to remove the "data" envelope.

File: WooCommerce/Classes/Authentication/AuthenticationManager.swift [hunk 2]
```diff
@@ -854,17 +862,85 @@ private extension AuthenticationManager {
...
+            guard let self else { return }
+            switch result {
+            case .success:
+                onSuccess()
+            case .failure(let error):
+                if case let RoleEligibilityError.insufficientRole(errorInfo) = error {
+                    self.showRoleErrorScreen(for: WooConstants.placeholderStoreID,
+                                             errorInfo: errorInfo,
+                                             in: navigationController,
+                                             onSuccess: onSuccess)
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/Authentication/AuthenticationManager.swift:895]: Unrelated to this PR:


Tapping `Retry` doesn't show a loading indicator.

**Review by selanthiraiyan — APPROVED** [review]
LGTM 🚢 

Good to merge after we address my comment about the mock JSON file `user-complete-without-data.json`.

File: WooCommerce/Classes/Authentication/AuthenticationManager.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/Authentication/AuthenticationManager.swift:895]: I added a quick fix in 0cd2149de6a277f947dde9609f2ea2f432f0a300 and 6b078f3c223ef8b2f540b12a683d4818519e6d2e. Here's a recording of the result:

https://user-images.githubusercontent.com/5533851/210319054-03a083c7-92fe-4a0e-b3a2-112354e86d0f.mp4

File: Networking/NetworkingTests/Responses/user-complete-without-data.json (hunk 1 above)
Comment (itsmeichigo) [reply Networking/NetworkingTests/Responses/user-complete-without-data.json]: Thanks for noticing! Fixed in 123eacb0892f576abae2621416917ed9d6bd5cd8.

**Pull request merged**